bpo-29951: Include function name for some error messages in PyArg_ParseTupleAndKeywords

MSeifert04 opened the pull request:
Issue: [29951](https://bugs.python.org/issue29951)

I wasn't sure if this needs to be tested and if there is a dedicated test-file for these functions.

MSeifert04 committed:
bpo-29951: Include function name for some error messages in PyArg_ParseTupleAndKeywords

serhiy-storchaka commented on Python/getargs.c:
```diff
@@ -1705,8 +1705,10 @@ vgetargskeywords(PyObject *args, PyObject *kwargs, const char *format,
             if (max < nargs) {
                 PyErr_Format(PyExc_TypeError,
-                             "Function takes %s %d positional arguments"
+                             "%s%s takes %s %d positional arguments"
```
Use `%.200s` instead of just `%s` for function name.

MSeifert04 committed:
bpo-29951: Include function name for PyArg_ParseTupleAndKeywords errors

Also changed format specifier for function name from "%s" to "%.200s"
and exception messages should start with lowercase letter

serhiy-storchaka commented on Lib/test/test_exceptions.py:
```diff
@@ -1089,7 +1089,7 @@ def test_attributes(self):
         self.assertEqual(exc.path, 'somepath')
 
-        msg = "'invalid' is an invalid keyword argument for this function"
+        msg = "'invalid' is an invalid keyword argument for ImportError()"
```
`msg` is a regular expression. It is matched only by accident. Either escape or remove `()`.

MSeifert04 replied:
Oups. Thanks for spotting this.

If escaped it also needs to be a raw string then, right?

serhiy-storchaka replied:
Yes, if you decide to escape `()`. But you can just remove ` for ImportError()`. This message can be changed in future again (for example to ` for ImportError constructor`).

serhiy-storchaka commented on Lib/test/test_getargs2.py:
```diff
@@ -626,16 +626,16 @@ def test_required_args(self):
         # required arg missing
         with self.assertRaisesRegex(TypeError,
-            r"Required argument 'required' \(pos 1\) not found"):
+            r"required argument for function 'required' \(pos 1\) not found"):
```
This looks as the function has the name "required". May be omit "for function" if function name is not specified?

MSeifert04 committed:
bpo-29951: Changed exception message for missing required argument

exarkun commented:
> @MSeifert04, thanks for your PR! By analyzing the history of the files in this pull request, we identified @serhiy-storchaka, @exarkun and @loewis to be potential reviewers.

Please remove me from the mention-bot configuration.

vadmium commented on Python/getargs.c:
```diff
@@ -1705,8 +1705,10 @@ vgetargskeywords(PyObject *args, PyObject *kwargs, const char *format,
-                             "Function takes %s %d positional arguments"
+                             "%.200s%s takes %s %d positional arguments"
                              " (%d given)",
+                             (fname == NULL) ? "function" : fname,
```
Any reason why you changed from uppercase Function to lowercase “function”?

It may be worth changing an exception message if you can make it less confusing or easier to read. But changing just the letter case does not seem worth the pain of breaking the old message.

serhiy-storchaka replied:
I suggested this for uniformity with other error messages.

vadmium replied:
Okay, I didn’t notice, and it’s not really a big deal. It just seemed a style preference thing to me.

serhiy-storchaka commented on Lib/test/test_extcall.py:
```diff
@@ -221,7 +221,7 @@
     Traceback (most recent call last):
       ...
-    TypeError: f() keywords must be strings
+    TypeError: f() keyword arguments must be strings
```
After you have changed this I see that the former wording was more correct. I think it would be better to change the error message in `PyArg_ValidateKeywordArguments()` rather than in all other code.

MSeifert04 committed:
bpo-29551: undo the "keyword argument" part of the change

serhiy-storchaka commented on Python/getargs.c:
```diff
@@ -1752,8 +1754,10 @@ vgetargskeywords(PyObject *args, PyObject *kwargs, const char *format,
                 else {
-                    PyErr_Format(PyExc_TypeError, "Required argument "
-                                 "'%s' (pos %d) not found",
+                    PyErr_Format(PyExc_TypeError,  "%.200s%s missing required "
```
Wouldn't be better to omit the word "function" if the function name is not specified? Just "missing required argument 'foo' (pos 2)". What are your thoughts? Is it correct English and does it look better?

MSeifert04 replied:
I know that almost all other error messages used "function" or "this function" as placeholder when there's no name. But, yes, I mostly included it for consistency reasons, I have no preference either way.

(Note: I'm non native speaker so I would have to guess about correctness and/or looks)

vadmium replied:
“Missing required argument ‘foo’ (pos 2)” is correct English.

Whether including “function” may help would depend on the context. It may help when it is not clear from the exception context that a function call is involved. On the other hand, omitting the word makes the message simpler when the context is already clear.

serhiy-storchaka commented on Python/getargs.c:
```diff
@@ -1802,8 +1808,10 @@ vgetargskeywords(PyObject *args, PyObject *kwargs, const char *format,
                 /* arg present in tuple and in dict */
                 PyErr_Format(PyExc_TypeError,
-                             "Argument given by name ('%s') "
+                             "argument for %.200s%s given by name ('%s') "
```
Wouldn't be better to omit the words "for function" if the function name is not specified? Just "argument given by name ('foo') and position (2)". What are your thoughts?

vadmium replied:
“For function” probably doesn’t add much value to the exception messsage in this case. But it does not seem very important to remove it either: it is a judgement call.

serhiy-storchaka reviewed (commented):
LGTM, but I have two style questions.

MSeifert04 commented:
Is there anything that needs to be done?

I'm not sure about the remaining comments. Should I revert the "capitalization" changes and should I remove the "function" stubs or are these "optional" comments?

serhiy-storchaka commented:
I don't know. @vadmium, what are your thoughts? The patch LGTM in current form and will LGTM with any of these modifications.

vadmium commented:
I think these questions are just a judgement call for whoever writes the code and whoever accepts it. If it were just me, I tend to start exception messages with a capital letter so that the message stands better on its own, but I would also respect the existing style, and avoid changing exception messages unless they are significantly better. So if you think the new messages are clearer or have a more consistent style, go for it.
